### DISCUSSION TITLE
jsonschema validation with golang

### DISCUSSION THREAD

**Discussion by humblenginr** [cue-lang/cue#1780] [Q&A]
I want to parse a jsonschema into a cue file. When I use the command-line client to do it, it works as expected. But I could not find  the right way to do it with golang. I did something like this 

```golang
import (
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/encoding/json"
	"cuelang.org/go/encoding/jsonschema"
)
       // schema is []byte
        jsonSchema, err := json.Extract("", schema)
	if err != nil {
		// handle
	}

	cueJsonSchemaExpr := cueCtx.BuildExpr(jsonSchema)
	if cueJsonSchemaExpr.Err() != nil {
		// handle error
	}
	extractedSchema, err := jsonschema.Extract(cueJsonSchemaExpr, &jsonschema.Config{
		PkgName: "myPkg",
		Strict:  true,
	})
	if err != nil {
		// returns error
	}
```

but it returns some errors. Am I doing it the right way ?

**Comment by folliehiyuki** [comment]
Hi! I just started learning CUE and I'm also curious about this. The current library is marked as *experimental*, but I want to use it. Having example code for this would be appreciated.

I use [mage](https://magefile.org) to run tasks, so I prefer doing **import jsonschema** / **validate generated jsonschema against written CUE files** with go library than invoking `cmd.Run` multiple times.

**Comment by verdverm** [comment]
You want to use `cue/load` package, which does (most of) the fancy things the CLI does

- https://pkg.go.dev/cuelang.org/go@v0.6.0/cue/load
- https://cuetorials.com/go-api/loading/

You might also need to look at the Orphaned files, like this:

- https://github.com/hofstadter-io/hof/blob/_dev/lib/runtime/load.go#L228

**Comment by folliehiyuki** [comment]
Thanks for the suggestion, but I don't think the problem OP mentioned is solved yet.

> You want to use `cue/load` package, which does (most of) the fancy things the CLI does

This would handle the loading of CUE/JSON files into Go program. Great! But, how do I transform this loaded data (in CUE format) into proper CUE schema? It definitely needs to be `encoding/jsonschema` package, but it doesn't seem to work as I expected it to be.

For example, this [github-workflow.json](https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/github-workflow.json) is the source JSON input. The task is to output a CUE schema file like this [github-workflow.cue](https://github.com/hoangnh-optimizely/aws-playground/blob/main/cue.mod/pkg/github.com/SchemaStore/schemastore/src/schemas/json/github-workflow.cue).

The minimal code (please ignore the dumb fmt.Println and no error check):
```go
	inputBytes, _ := os.ReadFile("./github-workflow.json")

	ctx := cuecontext.New()
	inputCue := ctx.CompileBytes(inputBytes)

	A, _ := jsonschema.Extract(inputCue, &jsonschema.Config{
		ID:      "https://json.schemastore.org/github-workflow.json",
		PkgName: "json",
	})

	outputCue := ctx.BuildFile(A)
	fmt.Println(outputCue)
```

It outputs this

```json
{
	on: "branch_protection_rule" | "check_run" | "check_suite" | "create" | "delete" | "deployment" | "deployment_status" | "discussion" | "discussion_comment" | "fork" | "gollum" | "issue_comment" | "issues" | "label" | "member" | "milestone" | "page_build" | "project" | "project_card" | "project_column" | "public" | "pull_request" | "pull_request_review" | "pull_request_review_comment" | "pull_request_target" | "push" | "registry_package" | "release" | "status" | "watch" | "workflow_call" | "workflow_dispatch" | "workflow_run" | "repository_dispatch" | ["branch_protection_rule" | "check_run" | "check_suite" | "create" | "delete" | "deployment" | "deployment_status" | "discussion" | "discussion_comment" | "fork" | "gollum" | "issue_comment" | "issues" | "label" | "member" | "milestone" | "page_build" | "project" | "project_card" | "project_column" | "public" | "pull_request" | "pull_request_review" | "pull_request_review_comment" | "pull_request_target" | "push" | "registry_package" | "release" | "status" | "watch" | "workflow_call" | "workflow_dispatch" | "workflow_run" | "repository_dispatch", ..."branch_protection_rule" | "check_run" | "check_suite" | "create" | "delete" | "deployment" | "deployment_status" | "discussion" | "discussion_comment" | "fork" | "gollum" | "issue_comment" | "issues" | "label" | "member" | "milestone" | "page_build" | "project" | "project_card" | "project_column" | "public" | "pull_request" | "pull_request_review" | "pull_request_review_comment" | "pull_request_target" | "push" | "registry_package" | "release" | "status" | "watch" | "workflow_call" | "workflow_dispatch" | "workflow_run" | "repository_dispatch"] | {}
	jobs: {}
}
```

instead of the proper CUE schema file I got from running `cue import`
